num/__private/signed_integer_macros.h

```diff
@@ -0,0 +1,445 @@
+  inline void operator&=(T r)& noexcept {                         \
+    primitive_value &= r.primitive_value;                         \
+  }                                                               \
+  /** sus::concepts::BitOrAssign<##T##> trait. */                 \
+  inline void operator|=(T r)& noexcept {                         \
+    primitive_value |= r.primitive_value;                         \
+  }                                                               \
+  /** sus::concepts::BitXorAssign<##T##> trait. */                \
+  inline void operator^=(T r)& noexcept {                         \
+    primitive_value ^= r.primitive_value;                         \
+  }                                                               \
+  /** sus::concepts::ShlAssign trait. */                          \
+  inline void operator<<=(/* TODO: u32 */ uint32_t s)& noexcept { \
+    ::sus::check(s < T::BITS());                                  \
+    primitive_value <<= s;                                        \
+  }                                                               \
+  /** sus::concepts::ShrAssign trait. */                          \
+  inline void operator>>=(/* TODO: u32 */ uint32_t s)& noexcept { \
+    ::sus::check(s < T::BITS());                                  \
+    primitive_value >>= s;                                        \
+  }                                                               \
+  static_assert(true)
+
+#define _sus__signed_abs(T)                                                    \
+  /** Computes the absolute value of itself.                                   \
+   *                                                                           \
+   * The absolute value of ##T##::MIN() cannot be represented as an ##T##, and \
+   * attempting to calculate it will panic.                                    \
+   */                                                                          \
+  constexpr inline T abs() const& noexcept {                                   \
+    ::sus::check(primitive_value != MIN_PRIMITIVE);                            \
+    if (primitive_value >= 0)                                                  \
+      return primitive_value;                                                  \
+    else                                                                       \
+      return -primitive_value;                                                 \
+  }                                                                            \
+                                                                               \
+  /** Checked absolute value. Computes `abs()`, returning None if the current  \
+   * value is MIN().                                                           \
+   */                                                                          \
+  constexpr Option<T> checked_abs() const& noexcept {                          \
```

Comment:
I have a slight bias toward

```
... Option<T> abs() ...
T unchecked_abs() ...
```

Reply:
This is worth talking about a bit. I am copying the shape of Rust i32 obviously, where the default panics if you go out of bounds.

unchecked_abs() would be an unsafe fn that does not panic.

but there is room for a panic_abs(). or to have checked_abs() panic instead of returning an option. I am not sure about making all abs() code deal with Options though, it feels like it's maybe okay to opt into handling bad things, and crash if you thought there isn't but there is.

like abs(10, -10) would require an unwrap then.

So given all that - do you have more thoughts on your bias toward abs() returning Option?

Reply:
I'm going to land this to keep moving forward, but that's not a comment on not wanting to hear your thoughts on this.